services/scheduler/scheduler.go

```diff
@@ -0,0 +1,330 @@
+	}
+
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	task, err := s.convertDBTask(dbTask)
+	if err != nil {
+		return err
+	}
+
+	_ = s.scheduler.RemoveByTag(id)
+
+	// Don't add it to scheduler, if it's disabled.
+	if dbTask.Disabled {
+		return nil
+	}
+
+	j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+	if !dbTask.StartAt.IsZero() {
+		j = j.StartAt(dbTask.StartAt)
+	}
+
+	fn := s.wrapTask(task, dbTask.ID)
+	if _, err := j.Tag(dbTask.ID).Do(fn); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Service) loadFromDB() error {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	disabled := false
+	dbTasks, err := models.FindScheduledTasks(s.db.Querier, models.ScheduledTasksFilter{
+		Disabled: &disabled,
+	})
+	if err != nil {
+		return err
+	}
+
+	s.scheduler.Clear()
+
+	s.jobsMx.Lock()
+	defer s.jobsMx.Unlock()
+
+	for _, dbTask := range dbTasks {
+		task, err := s.convertDBTask(dbTask)
+		if err != nil {
+			return err
+		}
+
+		fn := s.wrapTask(task, dbTask.ID)
+		j := s.scheduler.Cron(dbTask.CronExpression).SingletonMode()
+		if !dbTask.StartAt.IsZero() {
+			j = j.StartAt(dbTask.StartAt)
+		}
+		scheduleJob, err := j.Tag(dbTask.ID).Do(fn)
+		if err != nil {
+			return err
+		}
+
+		s.jobs[dbTask.ID] = scheduleJob
+	}
+
+	return nil
+}
+func (s *Service) wrapTask(task Task, id string) func() {
+	return func() {
+		var err error
+		l := s.l.WithFields(logrus.Fields{
+			"id":       id,
+			"taskType": task.Type(),
+		})
+		ctx, cancel := context.WithCancel(context.Background())
+
+		s.taskMx.Lock()
+		s.tasks[id] = cancel
+		s.taskMx.Unlock()
+
+		defer func() {
+			cancel()
+			s.taskMx.Lock()
+			delete(s.tasks, id)
+			s.taskMx.Unlock()
+		}()
+
+		t := time.Now()
+		l.Debug("Starting task")
+		_, err = models.ChangeScheduledTask(s.db.Querier, id, models.ChangeScheduledTaskParams{
+			Running: pointer.ToBool(true),
+		})
+
+		if err != nil {
+			l.Errorf("failed to change running state: %v", err)
+		}
+
+		taskErr := task.Run(ctx)
+		if taskErr != nil {
+			l.Error(taskErr)
+		}
+		l.WithField("duration", time.Since(t)).Debug("Ended task")
+
+		s.taskFinished(id, taskErr)
+	}
+}
+
+func (s *Service) taskFinished(id string, taskErr error) {
+	s.jobsMx.RLock()
+	job := s.jobs[id]
+	s.jobsMx.RUnlock()
+
+	l := s.l.WithField("id", id)
+
+	dbTask, err := models.FindScheduledTaskByID(s.db.Querier, id)
+	if err != nil {
+		return
+	}
+
+	params := models.ChangeScheduledTaskParams{
+		Running: pointer.ToBool(false),
+	}
+
+	if taskErr == nil {
+		params.Succeeded = pointer.ToUint(dbTask.Succeeded + 1)
+		params.Error = pointer.ToString("")
+	} else {
+		params.Failed = pointer.ToUint(dbTask.Failed + 1)
+		params.Error = pointer.ToString(taskErr.Error())
+	}
+
+	if job != nil {
+		nextRun := job.NextRun().UTC()
+		lastRun := job.LastRun().UTC()
+		params.NextRun = &nextRun
+		params.LastRun = &lastRun
+	} else {
+		l.Errorf("failed to find scheduled task: %v", err)
+	}
+
+	_, err = models.ChangeScheduledTask(s.db.Querier, id, params)
```

Comment:
I think transaction is need for `FindScheduledTaskByID` and `ChangeScheduledTask` calls.